### PR TITLE
WIP add compression to supend stream

### PR DESCRIPTION
**Pull request by lindig**
This does not work (yet) and I'm looking for faults with the approach.

### CODE REVIEW

File: ocaml/xenopsd/xc/domain.ml [hunk 1]
```diff
@@ -27,6 +27,14 @@ open D
 
 let finally = Xapi_stdext_pervasives.Pervasiveext.finally
 
+module Compression = struct
+  let scheme = "gzip"
```
Comment (edwintorok) [inline ocaml/xenopsd/xc/domain.ml:31]: `gzip` is quite slow for compression.
Would be useful to use something like `lz4` or `zstd -1` (there is already support for zstd in xapi-compress).
Lz4 is very fast (at the expense of slight reduction in compression ratio), and `zstd` is both faster and compresses better than gzip (depending on setting). `gzip` is only reasonably fast for decompression.
See benchmarks at https://lz4.github.io/lz4/ and https://facebook.github.io/zstd/#benchmarks

What matters here is compression time + network transfer time + decompression time. Although lets say compression and decompression runs on different machines and in parallel, so: `max(compression time, decompression time) + network transfer time`. Network transfer time will depend on compression ratio, so there is a tradeoff between time taken to compress a little bit more, or being a lot faster on compression and transferring a little bit more.
The lz4 benchmarks take into consideration the total transfer time: <2s for LZ4, vs >11s for zlib (gzip)

Comment (edwintorok) [reply ocaml/xenopsd/xc/domain.ml:31]: Quoting from the ztd benchmarks:



Compressor name | Ratio | Compression | Decompress.
-- | -- | -- | --
zlib 1.2.11 -1 | 2.743 | 90 MB/s | 400 MB/s
zstd 1.4.5 -1 | 2.884 | 500 MB/s | 1660 MB/s
zstd 1.4.5 --fast=1 | 2.434 | 570 MB/s | 2200 MB/s
zstd 1.4.5 --fast=3 | 2.312 | 640 MB/s | 2300 MB/s
zstd 1.4.5 --fast=5 | 2.178 | 700 MB/s | 2420 MB/s
lz4 1.9.2 | 2.101 | 740 MB/s | 4530 MB/s
```

File: ocaml/xenopsd/xc/domain.ml [hunk 2]
```diff
@@ -1314,13 +1322,36 @@ let restore_common (task : Xenops_task.task_handle) ~xc ~xs
...
+              return xr
+          | _ ->
+              error "Was expecting Xenops header but didn't find it" ;
+              Error Suspend_image_failure
+        )
+        |> function
+        | Ok header ->
+            header
+        | Error e ->
+            raise e
```
Comment (robhoes) [inline ocaml/xenopsd/xc/domain.ml:1352]: Think you need `cancel_on_error` as well (defined below).

**Comment by lindig** [comment]
This works in the sense that a stream with uncompressed header and gzip-compressed body is written and it appears the uncompression works in principle. The sizes reported for headers during writing and reading match. But there is some problem. `xenguesthelper` seems to have trouble with the data it receives and the decoding `gzip` exists with error code 1.

```
Mar 30 15:27:50 xrtuk-01-01 xenopsd-xc: [debug||17 |Async.VM.resume R:a94d5688cbce|mig64] Received response from emu-manager: xenguest 1044476 1044479
Mar 30 15:27:50 xrtuk-01-01 xenopsd-xc: [debug||17 |Async.VM.resume R:a94d5688cbce|mig64] Waiting for response from emu-manager
Mar 30 15:27:50 xrtuk-01-01 xenopsd-xc: [debug||66 |Async.VM.resume R:a94d5688cbce|mig64] Restored Libxc state
Mar 30 15:27:50 xrtuk-01-01 xenopsd-xc: [debug||66 |Async.VM.resume R:a94d5688cbce|mig64] Reading next header... (fd=35)
Mar 30 15:27:50 xrtuk-01-01 xenopsd-xc: [debug||66 |Async.VM.resume R:a94d5688cbce|mig64] Read Qemu_trad header (length=9938)
Mar 30 15:27:50 xrtuk-01-01 xenopsd-xc: [debug||66 |Async.VM.resume R:a94d5688cbce|xenops] VM = a2b6d77d-64a4-2732-066c-a528f52333a7; domid = 7; reading 9938 bytes from /var/run/xen/qemu-r
Mar 30 15:27:50 xrtuk-01-01 xenopsd-xc: [debug||66 |Async.VM.resume R:a94d5688cbce|mig64] Reading next header... (fd=35)
Mar 30 15:27:50 xrtuk-01-01 xenopsd-xc: [debug||66 |Async.VM.resume R:a94d5688cbce|mig64] Read suspend image footer
Mar 30 15:27:50 xrtuk-01-01 xenopsd-xc: [error||17 |Async.VM.resume R:a94d5688cbce|xenguesthelper] Memory F 25869252 KiB S 0 KiB T 32755 MiB
Mar 30 15:27:50 xrtuk-01-01 xenopsd-xc: [debug||17 |Async.VM.resume R:a94d5688cbce|mig64] Finished emu-manager result processing
Mar 30 15:27:50 xrtuk-01-01 xenopsd-xc: [debug||17 |Async.VM.resume R:a94d5688cbce|mig64] Reader threads completed successfully
Mar 30 15:27:50 xrtuk-01-01 xenopsd-xc: [debug||17 |Async.VM.resume R:a94d5688cbce|mig64] VM = a2b6d77d-64a4-2732-066c-a528f52333a7; domid = 7; store_mfn = 1044476; console_mfn = 1044479
Mar 30 15:27:50 xrtuk-01-01 xenopsd-xc: [error||17 |Async.VM.resume R:a94d5688cbce|xapi_compression] nice failed to decompress: exit code 1
```

**Comment by lindig** [comment]
```
sh-4.2# dd iflag=skip_bytes skip=1669 if=/dev/$DEVICE | gunzip > /dev/null
10776572+1 records in
10776572+1 records out
5517605243 bytes (5.5 GB) copied, 116.458 s, 47.4 MB/s
```

This tries to decompress the VDI by skipping over the header and it seems to work, suggesting the entire rest of the stream is compressed (as I had expected).

**Comment by lindig** [comment]
This is the end of the stream. According to `Xenops_record` the `End_of_image` is 
marked with `0xffffL`

```
sh-4.2# dd iflag=skip_bytes skip=1669 if=/dev/$DEVICE | gunzip | hexdump -C | tail -20
10776572+1 records in
10776572+1 records out
5517605243 bytes (5.5 GB) copied, 278.237 s, 19.8 MB/s
100881240  3f 3f 00 00 00 00 00 2a  00 2a 00 00 2a 2a 2a 00  |??.....*.*..***.|
100881250  00 2a 00 2a 2a 15 00 2a  2a 2a 15 15 15 15 15 3f  |.*.**..***.....?|
100881260  15 3f 15 15 3f 3f 3f 15  15 3f 15 3f 3f 3f 15 3f  |.?..???..?.???.?|
100881270  3f 3f 00 15 00 00 15 2a  00 3f 00 00 3f 2a 2a 15  |??.....*.?..?**.|
100881280  00 2a 15 2a 2a 3f 00 2a  3f 2a 00 15 15 00 15 3f  |.*.**?.*?*.....?|
100881290  00 3f 15 00 3f 3f 2a 15  15 2a 15 3f 2a 3f 15 2a  |.?..??*..*.?*?.*|
1008812a0  3f 3f 15 00 00 15 00 2a  15 2a 00 15 2a 2a 3f 00  |??.....*.*..**?.|
1008812b0  00 3f 00 2a 3f 2a 00 3f  2a 2a 15 00 15 15 00 3f  |.?.*?*.?**.....?|
1008812c0  15 2a 15 15 2a 3f 3f 00  15 3f 00 3f 3f 2a 15 3f  |.*..*??..?.??*.?|
1008812d0  2a 3f 15 15 00 15 15 2a  15 3f 00 15 3f 2a 3f 15  |*?.....*.?..?*?.|
1008812e0  00 3f 15 2a 3f 3f 00 3f  3f 2a 15 15 15 15 15 3f  |.?.*??.??*.....?|
1008812f0  15 3f 15 15 3f 3f 3f 15  15 3f 15 3f 3f 3f 15 3f  |.?..???..?.???.?|
100881300  3f 3f 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |??..............|
100881310  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
100881540  00 00 00 00 00 00 01 00  09 b0 c4 04 00 03 00 00  |................|
100881550  20 00 41 00 00 04 00 08  00 00 00 00 00 00 00 00  | .A.............|
100881560  00 00 00 10 00 00 00 00  7f 00 ff ff 00 00 00 00  |................|
100881570  00 00 00 00 00 00 00 00  00 00                    |..........|
10088157a
```

**Comment by lindig** [comment]
I think this approach is not going to work because we have too little control over what gzip is reading during uncompress.

**Pull request closed without merging**